### ISSUE TITLE
Runtime vs Eager not working for Gems in Rails application

### ISSUE THREAD

**Issue by kbaum** [danmayer/coverband#261]
When tracking gems in production, I see that overall coverage and runtime coverage are the same:

![image](https://user-images.githubusercontent.com/96786/57771281-44c4a280-76e0-11e9-9ef8-8d13f10187d1.png)

Looking at the details, I can see that loading the class definitions is marked as runtime:

![image](https://user-images.githubusercontent.com/96786/57771417-a97ffd00-76e0-11e9-8f4f-e6e8d0fd5861.png)

My guess is that bundler does not load gems as part of rails eager_load. Since our [current implementation](https://github.com/danmayer/coverband/blob/master/lib/coverband/utils/railtie.rb#L6) relies on rails eager_load to track eager_load coverage, gems are left out of this functionality.

Not sure of the right way to handle this, but being able to track gem usage at runtime would be a great feature as I think removing gems can dramatically reduce a rails process memory footprint and startup time.

We could attempt to capture bundle require as part of eager_load data? This might be confusing though as it is not technically part of rails eager_loading. This makes me think that we got the name wrong and perhaps what we want to capture is "initialization" coverage data vs "runtime".

**Comment by kbaum** [comment]
So I don't think my guess is correct. I was having difficulties reproducing locally even though this was happening in our heroku production environment. I cleared redis and set track_gems to true in production and I still saw the issue. I then cleared coverage again in production and restart one more time and the issue is no longer happening. Can't explain what I am seeing and it is difficult to trace down since I cannot reproduce consistently on my local machine.

**Comment by danmayer** [comment]
so another thing on this one was I know after the ignore first report I didn't see this so we might still be missing something on handling 'initialize' vs runtime that is just a bigger issue than eager_loading (as you have mentioned)

**Comment by kbaum** [comment]
So looks like this problem happens when using a rake task that depends on environment. This makes sense given what we've learned previously. Eager loading is disabled but application.rb requires all of the Gems:

```ruby
Bundler.require(*Rails.groups)
```

This is essentially an eager load of all of those files on initialization. To address, I added this within application.rb:

```ruby
require 'coverband'
Coverband.eager_loading_coverage!
Bundler.require(*Rails.groups)
Coverband.report_coverage
Coverband.runtime_coverage!
```

Thinking about how we would address this issue for everyone else and might be difficult given `Bundle.require(*Rails.groups)` happens before coverband is required. This may require users to modify their application.rb.
